Fix - app deployment issue

LalitDeore opened the pull request:
@frikky Please review once before merging
1.  Fix specical character in issue in function name
2. Fix same parameter pass two time in same function issue.

0x0elliot commented on codegen.go:
```diff
@@ -747,14 +747,55 @@ func MakePythoncode(swagger *openapi3.Swagger, name, url, method string, paramet
+	// Create a map to track used parameters
+	usedParams := make(map[string]bool)
+
+	addParameter := func(param string) string {
```
make this an independent function for cleanliness?

0x0elliot commented on codegen.go:
```diff
@@ -2167,6 +2208,10 @@ func FixFunctionName(functionName, actualPath string, lowercase bool) string {
 	functionName = strings.Replace(functionName, " ", "_", -1)
 	functionName = strings.Replace(functionName, "-", "_", -1)
 
+	functionName = strings.Replace(functionName, "%", "", -1)
```
can you try handling this by whitelisting characters instead of blacklisting them? notice how:
-  " " -> "_" and "-" -> "_"
- but, the rest are just stripped. we can't keep dealing with bullshit characters. might as well strip them away and start from a predictable charset (a-z, A-Z, 1-9) 

@frikky any other reviews?

frikky commented on codegen.go:
```diff
@@ -747,14 +747,55 @@ func MakePythoncode(swagger *openapi3.Swagger, name, url, method string, paramet
-		queryString,
-		bodyParameter,
-		verifyParam,
+		addParameter(strings.TrimPrefix(authenticationParameter, ",")),
```
I'm scared of this. Sounds like it would break literally everything.

frikky commented:
In short: Your code is not at all trustworthy for our problem, unless it's tested across 1000+ apps and doesn't break a single one. I'm scared of deploying it due to it being VERY big changes to the actual function definitions.

This does not feel like a good solution, as it's a solution on the code generation side (python) and not a solution on the control-mechanism of the incoming OpenAPI

LalitDeore commented:
This two commit ( [commit 1](https://github.com/Shuffle/shuffle-shared/pull/154/commits/703808b7fe6749a288e793a4836db7a05c7e54e0), [commit 2](https://github.com/Shuffle/shuffle-shared/pull/154/commits/5eeae65ccfefe2223f74f5a9cde3d2f58c74e611) ) fixes the following issues:

1. Created a new function to check if the same parameter is passed multiple times to the same function without using a map.
2. Added a function to validate the URL format, ensuring that incorrect URLs like `url}/api/v1/user/{id/new` are corrected to `{url}/api/v1/user/{id}/new`
3. Implemented a regex validation to check that Python function names can only contain letters (a-z, A-Z) and numbers (0-9), with no special characters and replace special character with "_". 

Most app deployments were failing due to these issues. To validate this changes, I deployed 8–10 different apps on Frankfurt, and it worked successfully.

frikky commented:
As this is related to App Health, I'm assigning it to you @yashsinghcodes. You are testing similar stuff already, and I need you to make sure we can push this code without it causing even a single old app to stop functioning.

Changing names of actions in one place (python) but not another (app.yaml) have a tendency to create chaos.

yashsinghcodes commented:
Hey @LalitDeore

Found some issue with the function `HandleCheckValidURL` it is creating a problem, when I was testing here is the error I was getting when deploying a new app (cloud function).

<img width="1462" alt="Screenshot 2025-03-12 at 12 54 59 AM" src="https://github.com/user-attachments/assets/b3b1e12e-42dc-4949-a07f-4e6040bf9aff" />


Here is the app id of the app I used: a286d101b679672403456f056a4c600f

feel free to let me know if you want any information on it.

frikky commented:
> Hey @LalitDeore
> 
> Found some issue with the function `HandleCheckValidURL` it is creating a problem, when I was testing here is the error I was getting when deploying a new app (cloud function).
> 
> <img alt="Screenshot 2025-03-12 at 12 54 59 AM" width="1462" src="https://private-user-images.githubusercontent.com/32360914/421541589-b3b1e12e-42dc-4949-a07f-4e6040bf9aff.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyOTM1NzAsIm5iZiI6MTc0MjI5MzI3MCwicGF0aCI6Ii8zMjM2MDkxNC80MjE1NDE1ODktYjNiMWUxMmUtNDJkYy00OTQ5LWEwN2YtNGU2MDQwYmY5YWZmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE4VDEwMjExMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZlYmZhOGFkYmIwMWNjOGQwZWU1MTQ4ZTEyYWQ2NDQwMDExNTVhNTVhODgyNDcwYzAzYjFhZDJjZGQ0ODNhYmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.yN4f7L_eDOY3T1c9nNEdZr985IAhfqzYxNwxe2idbWU">
> Here is the app id of the app I used: a286d101b679672403456f056a4c600f
> 
> feel free to let me know if you want any information on it.

The TWO things I need to see from this PR to be confident in merging it:
1. There NEEDS to be a sync between the Python function name and the Api.yaml function. Right now, you are ONLY fixing the actual Python code, but leaving out the user experience entirely.  
2. It needs to solve literally 100% of the current random code generator issues we have. Right now, I can still see a ton of cloud functions that have errors. If this PR works, they should not exist anymore.

frikky commented:
Closing as you haven't sent any feedback in a while. 

This kind of PR is VERY scary to merge without proof of it not breaking app deployments.
